content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md

```diff
@@ -0,0 +1,75 @@
++++
+title = "Peephole Improvements for Numpy with JAX"
+[extra]
+latex = true
+[[extra.authors]]
+name = "Aaron Tucker"
+link = "https://www.cs.cornell.edu/~aarondtucker/"
++++
+
+# Goal
+Sometimes when writing code it is possible to run into numerical stability issues. These issues are typically dealt with through a mix of thoughtful fixes, painful debugging, or knowing various "tricks". What if your code did the tricks for you? This project's goal is to use "peephole improvements" to make it easier to write numerically stable code, rather than the end-user needing to know about them beforehand. Peephole improvements are based on peephole optimizations, but instead of trying to replace small parts of code with equivalent faster code they replace small parts of code with similar "better code" -- improving numerical stability changes the behavior of the program by avoiding NaNs in situations where the original wouldn't, but that is typically appreciated as long as it doesn't change other behavior too much.
+
+# Implementation
+The implementation of the project builds on several ideas explored in class, such as peephole optimizations and intermediate representations.
+
+## Design
+The basic design of the project is to convert code which uses numpy into an intermediate representation, perform the peephole optimizations on the intermediate representation, and then execute the code.
```

Comment:
I'd probably spell it NumPy, and maybe link to [its homepage](https://numpy.org).